### ISSUE TITLE
Add a SlotFill to Toolbar in the header

### ISSUE THREAD

**Issue by matthewbrent** [WordPress/gutenberg#16988]
**Is your feature request related to a problem? Please describe.**
Currently Several SlotFills exist in the editor application. Some are wrapped as plugins and others are not. It is currently possible to add a Fill/Plugin to the right hand side of the toolbar, next to where the inspector toggle is. There are none for the left hand side of the toolbar where the add block button is.

Currently the Block Menu can be added to this area by enabling the Top Toolbar setting in the more tools and options ellipsis menu in the top right however, after reading the source, this appears to be added to the toolbar using a boolean flag rather than the SlotFill
[see the line here](https://github.com/WordPress/gutenberg/blob/c7faf14771ebd2374830cb061e703a42a577b641/packages/edit-post/src/components/header/header-toolbar/index.js#L43)

**Describe the solution you'd like**
Refactor the Header Toolbar to replace the conditional logic that renders the BlockToolbar with a SlotFill that will both make sense from a refactor and also make it possible to add more Fills to this location (and also allow a plugin for this location at a later date)

**Comment by matthewbrent** [comment]
I'm stacked right now, but once I get some time I'd be willing to make a PR if that would help?

**Comment by lukecarbis** [comment]
It would be nice to have a slot fill for:

- Header Left
    - Inserter
    - Undo
    - Redo
    - Table of contents
    - Navigation
- Header Center
- Header Right
    - Post publish
    - Settings
    - More menu

**Comment by jeffpaul** [comment]
As further background on the desire for a solution like is proposed in this issue, the following plugins are examples that currently integrate functionality into the admin bar:
- [Site Kit by Google](https://wordpress.org/plugins/google-site-kit/): 400,000+ active installs and 3,054,354 downloads
- [Query Monitor](https://wordpress.org/plugins/query-monitor/): 100,000+ active installs and 2,285,090 downloads
- [Debug Bar](https://wordpress.org/plugins/debug-bar/): 30,000+ active installs and 845,770 downloads
- [Toolbar Extras for Elementor](https://wordpress.org/plugins/toolbar-extras/): 20,000+ active installs and 219,797 downloads
- [Admin Bar User Switching](https://wordpress.org/plugins/admin-bar-user-switching/): 2,000+ active installs and 22,450 downloads
- [Distributor](https://github.com/10up/distributor): _disclaimer that I personally oversee this plugin as part of my job_

There are also newly released plugins that are integrating with the admin bar (e.g., [Accessible Content from CampusPress](https://campuspress.com/free-accessibility-plugins-themes-for-wordpress/)) so this is an issue that will continue to expand in impact as other plugins integrate with the admin bar without an obvious replacement in Gutenberg's fullscreen mode.

I'll also note that there are plugins that are finding other ways to integrate with Gutenberg's fullscreen mode in lieu of the admin bar (e.g., [Yoast](https://wordpress.org/plugins/wordpress-seo/) integrating into both the sidebar and below the content editor), but without a standard recommendation from the Gutenberg team on how to pivot from the admin bar into a Gutenberg-first world we'll likely see a wide variety of integrations that create a disjointed and confusing experience.

So while I can appreciate the desire to hold off on opening up a slotfill in Gutenberg's header to allow for time to consider all options, we must also consider that in lieu of a slotfill that we provide a best practice recommendation to plugins that currently integrate with the admin bar on how to replicating that functionality into Gutenberg.  Making this sort of recommendation will go a long way in ensuring a standard UX for users expectation for plugins functionality within Gutenberg.

**Comment by gziolo** [comment]
Let me reshare first the latest proposal from @ryanwelcher:

> I believe we should be more purposeful about what is placed into Slot locations. Generally speaking, opening a spot to place anything is going to lead to very bad user experience and bad times overall.
> 
> This new location should be restricted to the same type of controls that exist now.  I propose that the Fill return a `ToolbarItem` that set to use a custom `DropDown` using a `Button` with `isPrimary` set to false.
> 
> In addition to the `className`, `classContentName` and `position` related props, the fill will accept a prop for the `icon` to display and a `renderContent` prop that will be passed to the inner DropDown component. This will keep each item added here standardized and prevent abuse for this Slot.
> 
> **Example implementation:**
> ```jsx
> const render = () => (
>     <PluginHeaderToolbar
>         className="my-plugin"
>         classContentName="my-plugin-content"
>         position="bottom left"
>         renderContent={() => <div>Content to render</div>}
>       />
> );
> registerPlugin("my-plugin", { render, icon: "twitter" });
> ```
> **Screenshot:**
> <img width="1173" alt="plugin-header-toolbar" src="https://user-images.githubusercontent.com/1259027/84717628-5bbd7000-af44-11ea-863c-63b0e98a8dc9.png">
> 
> **Screenshot with Block Controls in the Toolbar:**
> <img width="1143" alt="plugin-header-toolbar-block-toolbar" src="https://user-images.githubusercontent.com/1259027/84717744-c2db2480-af44-11ea-9dff-9de3e10b0581.png">
> 
> In addition to the changes to SlotFill, I also propose that this slot is only rendered when `isLargeViewport` is true. These items should be considered secondary and treated similarly to the ToolSelector as they are not crucial to content editing.

**Comment by gziolo** [comment]
> As further background on the desire for a solution like is proposed in this issue, the following plugins are examples that currently integrate functionality into the admin bar

I think it's rather an issue that was created when the full-screen mode was set as the default mode. I know that there are some explorations that seek to make it easier to make the admin bar visible. Anyway, it's only slightly related because it's hard to tell how many of those integrations make sense in the context of the editing content on a currently open page.

> I'll also note that there are plugins that are finding other ways to integrate with Gutenberg's fullscreen mode in lieu of the admin bar (e.g., Yoast integrating into both the sidebar and below the content editor), but without a standard recommendation from the Gutenberg team on how to pivot from the admin bar into a Gutenberg-first world we'll likely see a wide variety of integrations that create a disjointed and confusing experience.

I think Yoast is a good example of how integration should look like. They were involved in the development process and largely influenced what is possible today. They use `PluginSidebar` component that is integrated with a menu item in the More Menu and a button in the Pinned Items (on the right). In the footer, they integrate the old interface that was based on meta boxes.

I understand the need for more integrations and I personally discussed many times some ideas for SlotFill based plugin:
- menu item in More Menu + button in the pinned item that opens a modal dialog – similar to the existing Copy All Content 
- menu item in More Menu + button in the pinned item that executes an action – similar to Block Manager or Keyboard Shortcuts
- there was also this proposal from @jasmussen to have full-screen takeover plugin

I would like to understand what types of extensions  are missing before we go into the specifics where to put them. What @ryanwelcher proposed is quite close to what More Menu is, it's just in a different place. I'm not against making it possible to put plugin integrations in other places, but maybe we could use one API and only introduce a new prop in the existing plugin components that would allow developers to render them in a different context.

**Comment by JustinSainton** [comment]
Adding an additional voice in here for allowing for SlotFill patterns used more globally. As others have already surfaced - some of the bigger pain points our publishing clients are experiencing in switching from the classic editor to Gutenberg include having top-level, global controls available in a first-class slot, right, where they'd expect them. 

Especially what other have called the "Header Left" section - not having that be a fillable slot seems like a significant oversight.

**Comment by PypWalters** [comment]
The AMP plugin currently adds a button to the right of the preview button
![Screen Shot 2021-09-15 at 9 07 52 AM](https://user-images.githubusercontent.com/23455618/133627314-95ad0373-cb72-4e67-a252-2a912af3eab3.png)

This is their solution for that: https://github.com/ampproject/amp-wp/blob/81b9df742d9f78fb5ee3657622d21eb8c7926f64/assets/src/block-editor/plugins/wrapped-amp-preview-button.js

**Comment by gziolo** [comment]
> The AMP plugin currently adds a button to the right of the preview button
> ![Screen Shot 2021-09-15 at 9 07 52 AM](https://user-images.githubusercontent.com/23455618/133627314-95ad0373-cb72-4e67-a252-2a912af3eab3.png)
> 
> This is their solution for that: https://github.com/ampproject/amp-wp/blob/81b9df742d9f78fb5ee3657622d21eb8c7926f64/assets/src/block-editor/plugins/wrapped-amp-preview-button.js

There is a very promising PR #31984 from @delawski that makes the Preview dropdown extensible. In the future, the AMP plugin could use a more formal approach for preview when the functionality is available.

Anyway, that you for sharing the use case and the approach that plugins currently take.

**Comment by lukecarbis** [comment]
Just want to bump this issue. I've come across the need to add buttons to the toolbar many times since I commented in 2019. AMP's DOM manipulation solution is too hacky (sorry @westonruter and @pierlon) to be a permanent solution.

@ryanwelcher's PR may need a refresh, but it looks like a good approach to me.

**Comment by westonruter** [comment]
> AMP's DOM manipulation solution is too hacky

Agreed. And the hacks continue: https://github.com/ampproject/amp-wp/pull/6881#issuecomment-1029245573

**Comment by richtabor** [comment]
We're doing something similar at Extendify to add a button to the toolbar. Elementor does itself as well, Stackable — quite a few others too. 

I think we should perhaps have a slot that can be used instead to clean up these additions, reduce conflicts, and also build in a method for users to turn toolbar items off/on via the Preferences modal (like we have for the block manager perhaps). 

A rough-cut visual:

<img width="487" alt="CleanShot 2022-02-22 at 08 59 07@2x" src="https://user-images.githubusercontent.com/1813435/155147196-29d185c8-45d5-4c48-a3f7-995685f63981.png">

**Comment by iandunn** [comment]
To clarify, it seems like there's several things discussed here that are related, but somewhat independent of each other:

* Add a Slot for `edit-post-header-toolbar__left`
* Add a Slot for the center area between `edit-post-header-toolbar__left` and `edit-post-header__settings`
* Add a Slot for `edit-post-header__settings`
* Admin bar customizations are hidden in fullscreen mode, some of which are related to post editing (and some aren't)

It seems like #17198 only addresses `edit-post-header-toolbar__left`.

**Comment by iandunn** [comment]
My use case is that I'd like to add a new button to the `edit-post-header__settings`

<table>
<tr>
<td>

![Screen Shot 2022-03-01 at 10 04 18 AM](https://user-images.githubusercontent.com/484068/156223927-809653dc-8251-4479-9367-4cb0d454e158.png)

</td>
</tr>
</table>

It's similar to Jeff's Distributor plugin, where a multisite has a canonical version of a post, and other sites "fork" it, but sometimes it's desirable to sync them.

It seems like the current recommendation is to follow what Yoast/Jetpack do, but I feel like that has several problems:

* It's hard to communicate what the button does in an icon
* The icon won't be directly next to the `Update` button, which is the most logical place for it
* An entire sidebar just for a button feels like overkill
* The user has to click 3 times to perform the action, instead of once (open sidebar, press button, then press `gear` icon to restore the normal UI)
* It'd be hidden on mobile screens

**Comment by iandunn** [comment]
It seems like one of the blockers to this and other new slots is [the lack of room in the UI](https://github.com/WordPress/gutenberg/pull/17198#issuecomment-643095880). I don't understand why that has to be a problem, though.

Could we adapt the mobile layout to the device and the fills, rather than trying to make it use the desktop layout? For example, the "left", "center", and "right" areas of the toolbar could each have their own row, and then we wouldn't have to hide features from users, or block extensibility.

<img width="453" alt="Screen Shot 2022-03-01 at 10 51 12 AM" src="https://user-images.githubusercontent.com/484068/156230582-8553e71e-ba3e-4ad0-a0a2-38aeb9c8fde2.png">

Are there tradeoffs that aren't obvious to me, or something else I'm missing?

**Comment by UranymusDev** [comment]
Any Updates on this? currently having to solve this exact issue and i am not sure if i should do it myself or just wait

**Comment by karmatosed** [comment]
This seems assigned to you @ryanwelcher but also it is flagged for 'needs design'. I am trying to move it forward a little or at least find the next step. I am wondering considering interface changes if this is even something needs design today or can we use options / some other existing interface?

**Comment by Mamaduka** [comment]
I would like to quote @youknowriad on this:

> A slot in the editor header is not a very semantic API. It’s unclear what it will be used for. This will prevent any redesign and iteration to the header area of the editor.

Ref: https://make.wordpress.org/core/2024/09/12/gutenberg-development-practices-and-common-pitfalls/.

Adding a SlotFill isn't technically a problem, but non-semantic APIs like it are often abused, resulting in bad UI/UX.

**Comment by PypWalters** [comment]
I totally agree that it could be abused, but as is demonstrated in the thread above, it is currently not impossible to hack something into that space, so if someone really wants to do it, they currently can. Maybe a slotfill is not the right answer because it creates too many possibilities. Maybe it should look like an ability to pass an icon, a tooltip and a callback, and then the button is created and managed by the block editor from there. In the preferences popup, users can then manage the visibility of the buttons in the toolbar just like they can currently manage document settings and blocks. See current preference popup below that I am referencing. There can also be a filter that developers can use to deactivate these buttons in the code base.

![Image](https://github.com/user-attachments/assets/07d2798c-412d-4958-a2b7-b0447a06ceb9)

If a button is deactivated (or maybe if we are in a mobile view), the button can move to a fallback component on the right sidebar or among the meta boxes. Maybe developers can even pass a fallback callback option when creating the button.

If the only solution for developers will be to hack a solution, then the users will have no hope of managing visibility. If we can create some kind of extendability for this then we can lock it down to ensure a good UI/UX, and consider user and developer preferences.

**Comment by widoz** [comment]
It might be a stupid idea, but have you considered the possibility of a secondary Header, below the main one. It could expand or collapse, being compact and allow icon mode or text mode, similar to what applications for general editing purpose are implementing. Having such an additional space, could make a refactoring easier if it needs to be moved somewhere else or redesigned. 

The header seems to be a critical place for extentions, preventing more editing contextual elements to have their space which would also be a problem for the user, mixing up different elements belonging to different scopes...

But probably would make also sense to understand how the extentions would render their content. For instance, the element in this header would mostly be an action triggering the opening of a menu, a modal or could also open an additional Complementary Area?

Regarding the Complementary Area, there could be the possibility to have a Plugin registering a Complementary Area with a Navigation Screen. Extensions could use a small abstraction to introduce their content. So that, if necessary in the future it would be possible to perform a refactoring, moving the implementation in another place.

Maybe (and I'm thinking outloud), the problem to solve right now is how to introduce an extensible concept which abstract the way and the place these elements are rendered, so that depending on how the editor evolve, it would still be possible to keep it coherent with the rest of the editor changes.

**Comment by Mamaduka** [comment]
An interesting idea, @widoz. A secondary toolbar could work with a toggle after the "Document Overview" button. If I remember correctly, the "Top Toolbar" used to render in that manner.

@mtias shared a good prerequisite for this API - https://github.com/WordPress/gutenberg/pull/52261#issuecomment-2323054979.

Note: Multiple toolbars may be displayed on smaller screens.

P.S. Let's avoid repeating the [toolbar mistakes](https://i.sstatic.net/pOAAU.jpg) from IE6 😄 

## Screenshot

<img width="578" height="784" alt="Image" src="https://github.com/user-attachments/assets/be85ba18-3ba9-44ef-b5d0-20db96f5c80b" />

**Comment by widoz** [comment]
Maybe the secondary toolbar could scroll horizontally? Like a slider with a drag&drop reordering possible?
I guess the alternative would be to go with predefined categories, thought the identification of them could be a tricky task.
